File: packages/program/programs/pact-insurance/src/state.rs
use anchor_lang::prelude::*;
use crate::constants::{MAX_AGENT_ID_LEN, MAX_CALL_ID_LEN, MAX_HOSTNAME_LEN};

#[account]
#[derive(InitSpace)]
pub struct ProtocolConfig {
    pub authority: Pubkey,
    pub treasury: Pubkey,
    pub usdc_mint: Pubkey,

    pub protocol_fee_bps: u16,
    pub min_pool_deposit: u64,
    pub default_insurance_rate_bps: u16,
    pub default_max_coverage_per_call: u64,
    pub min_premium_bps: u16,

    pub withdrawal_cooldown_seconds: i64,
    pub aggregate_cap_bps: u16,
    pub aggregate_cap_window_seconds: i64,

    pub claim_window_seconds: i64,
    pub max_claims_per_batch: u8,

    pub paused: bool,

    pub bump: u8,
}

impl ProtocolConfig {
    pub const SEED: &'static [u8] = b"protocol";
}

#[account]
#[derive(InitSpace)]
pub struct CoveragePool {
    pub authority: Pubkey,
    #[max_len(MAX_HOSTNAME_LEN)]
    pub provider_hostname: String,
    pub usdc_mint: Pubkey,
    pub vault: Pubkey,

    pub total_deposited: u64,
    pub total_available: u64,
    pub total_premiums_earned: u64,
    pub total_claims_paid: u64,

    pub insurance_rate_bps: u16,
    pub min_premium_bps: u16,
    pub max_coverage_per_call: u64,

    pub active_policies: u32,

    pub payouts_this_window: u64,
    pub window_start: i64,

    pub created_at: i64,
    pub updated_at: i64,
    pub bump: u8,
}

impl CoveragePool {
    pub const SEED_PREFIX: &'static [u8] = b"pool";
    pub const VAULT_SEED_PREFIX: &'static [u8] = b"vault";
}

#[account]
#[derive(InitSpace)]
pub struct UnderwriterPosition {
    pub pool: Pubkey,
    pub underwriter: Pubkey,
    pub deposited: u64,
    pub earned_premiums: u64,
    pub losses_absorbed: u64,
    pub deposit_timestamp: i64,
    pub last_claim_timestamp: i64,
    pub bump: u8,
}

impl UnderwriterPosition {
    pub const SEED_PREFIX: &'static [u8] = b"position";
}

#[account]
#[derive(InitSpace)]
pub struct Policy {
    pub agent: Pubkey,
    pub pool: Pubkey,
    #[max_len(MAX_AGENT_ID_LEN)]
    pub agent_id: String,
    pub agent_token_account: Pubkey,
    pub total_premiums_paid: u64,
    pub total_claims_received: u64,
    pub calls_covered: u64,
    pub active: bool,
    pub created_at: i64,
    pub expires_at: i64,
    pub bump: u8,
}

impl Policy {
    pub const SEED_PREFIX: &'static [u8] = b"policy";
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace)]
pub enum TriggerType {
    Timeout,
    Error,
    SchemaMismatch,
    LatencySla,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace)]
pub enum ClaimStatus {
    Pending,
    Approved,
    Rejected,
}

#[account]
#[derive(InitSpace)]
pub struct Claim {
    pub policy: Pubkey,
    pub pool: Pubkey,
    pub agent: Pubkey,
    #[max_len(MAX_CALL_ID_LEN)]
    pub call_id: String,
    pub trigger_type: TriggerType,
    pub evidence_hash: [u8; 32],
    pub call_timestamp: i64,
    pub latency_ms: u32,
    pub status_code: u16,
    pub payment_amount: u64,
    pub refund_amount: u64,
    pub status: ClaimStatus,
    pub created_at: i64,
    pub resolved_at: i64,
    pub bump: u8,
}

impl Claim {
    pub const SEED_PREFIX: &'static [u8] = b"claim";
}

File: packages/protocol-v2-client/src/state.ts
/**
 * Account decoders for `@q3labs/pact-protocol-v2-client`.
 *
 * Mirrors `packages/program/programs-pinocchio/pact-network-v2-pinocchio/src/state.rs`
 * bit-for-bit. Each struct has a 1-byte discriminator + 7 bytes of padding,
 * with the first domain field at offset 8 (enforced on the Rust side by
 * compile-time `const _: () = assert!(offset_of!(…) == 8)`). Trailing
 * `reserved: [u8; 64]` is intentionally omitted from the decoded TS type —
 * callers should not see implementation pad.
 *
 * Field offsets are derived from the Rust struct ordering plus `repr(C)`
 * natural alignment with zero implicit padding (the structs are laid out
 * specifically so that `bytemuck::Pod` accepts them without filler).
 */
import {
  makeReader,
  readBytes32,
  readI64LE,
  readPubkey,
  readU16LE,
  readU32LE,
  readU64LE,
  readU8,
  readUtf8Slice,
} from "./decoders.js";

/** Canonical base58-encoded Solana pubkey shape. Matches V1's type alias. */
export type Pubkey = string;

// ---------------------------------------------------------------------------
// LEN constants — total serialized size of each account.
// ---------------------------------------------------------------------------
export const PROTOCOL_CONFIG_LEN = 256;
export const COVERAGE_POOL_LEN = 320;
export const UNDERWRITER_POSITION_LEN = 184;
export const POLICY_LEN = 320;
export const CLAIM_LEN = 288;

// ---------------------------------------------------------------------------
// Discriminator bytes (account types, NOT instructions — these are stored at
// offset 0 of every account data buffer).
// ---------------------------------------------------------------------------
export const ACCOUNT_DISC_PROTOCOL_CONFIG = 0;
export const ACCOUNT_DISC_COVERAGE_POOL = 1;
export const ACCOUNT_DISC_UNDERWRITER_POSITION = 2;
export const ACCOUNT_DISC_POLICY = 3;
export const ACCOUNT_DISC_CLAIM = 4;

// ---------------------------------------------------------------------------
// Enums — mirror Rust `#[repr(u8)]` enums.
// ---------------------------------------------------------------------------
export enum TriggerType {
  Timeout = 0,
  Error = 1,
  SchemaMismatch = 2,
  LatencySla = 3,
}

export enum ClaimStatus {
  Pending = 0,
  Approved = 1,
  Rejected = 2,
}

// ---------------------------------------------------------------------------
// ProtocolConfig — discriminator 0, 256 bytes.
// ---------------------------------------------------------------------------
export interface ProtocolConfig {
  authority: Pubkey;
  oracle: Pubkey;
  treasury: Pubkey;
  usdcMint: Pubkey;
  minPoolDeposit: bigint;
  defaultMaxCoveragePerCall: bigint;
  withdrawalCooldownSeconds: bigint;
  aggregateCapWindowSeconds: bigint;
  claimWindowSeconds: bigint;
  protocolFeeBps: number;
  defaultInsuranceRateBps: number;
  minPremiumBps: number;
  aggregateCapBps: number;
  maxClaimsPerBatch: number;
  paused: number;
  bump: number;
}

export function decodeProtocolConfig(
  data: Buffer | Uint8Array
): ProtocolConfig {
  const v = makeReader(
    data,
    PROTOCOL_CONFIG_LEN,
    ACCOUNT_DISC_PROTOCOL_CONFIG,
    "ProtocolConfig"
  );
  return {
    authority: readPubkey(v, 8),
    oracle: readPubkey(v, 40),
    treasury: readPubkey(v, 72),
    usdcMint: readPubkey(v, 104),
    minPoolDeposit: readU64LE(v, 136),
    defaultMaxCoveragePerCall: readU64LE(v, 144),
    withdrawalCooldownSeconds: readI64LE(v, 152),
    aggregateCapWindowSeconds: readI64LE(v, 160),
    claimWindowSeconds: readI64LE(v, 168),
    protocolFeeBps: readU16LE(v, 176),
    defaultInsuranceRateBps: readU16LE(v, 178),
    minPremiumBps: readU16LE(v, 180),
    aggregateCapBps: readU16LE(v, 182),
    maxClaimsPerBatch: readU8(v, 184),
    paused: readU8(v, 185),
    bump: readU8(v, 186),
  };
}

// ---------------------------------------------------------------------------
// CoveragePool — discriminator 1, 320 bytes.
//
// `vaultBump` is the WP-8 cache stored in `_pad_tail[0]` (byte 250). Exposed
// even though no current builder consumes it because clients building
// signer-seeds for pool-PDA-signed CPIs need to know it without re-deriving.
// ---------------------------------------------------------------------------
export interface CoveragePool {
  authority: Pubkey;
  usdcMint: Pubkey;
  vault: Pubkey;
  providerHostname: string;
  providerHostnameLen: number;
  totalDeposited: bigint;
  totalAvailable: bigint;
  totalPremiumsEarned: bigint;
  totalClaimsPaid: bigint;
  maxCoveragePerCall: bigint;
  payoutsThisWindow: bigint;
  windowStart: bigint;
  createdAt: bigint;
  updatedAt: bigint;
  activePolicies: number;
  insuranceRateBps: number;
  minPremiumBps: number;
  bump: number;
  vaultBump: number;
}

export function decodeCoveragePool(data: Buffer | Uint8Array): CoveragePool {
  const v = makeReader(
    data,
    COVERAGE_POOL_LEN,
    ACCOUNT_DISC_COVERAGE_POOL,
    "CoveragePool"
  );
  const providerHostnameLen = readU8(v, 248);
  return {
    authority: readPubkey(v, 8),
    usdcMint: readPubkey(v, 40),
    vault: readPubkey(v, 72),
    providerHostname: readUtf8Slice(v, 104, 64, providerHostnameLen),
    providerHostnameLen,
    totalDeposited: readU64LE(v, 168),
    totalAvailable: readU64LE(v, 176),
    totalPremiumsEarned: readU64LE(v, 184),
    totalClaimsPaid: readU64LE(v, 192),
    maxCoveragePerCall: readU64LE(v, 200),
    payoutsThisWindow: readU64LE(v, 208),
    windowStart: readI64LE(v, 216),
    createdAt: readI64LE(v, 224),
    updatedAt: readI64LE(v, 232),
    activePolicies: readU32LE(v, 240),
    insuranceRateBps: readU16LE(v, 244),
    minPremiumBps: readU16LE(v, 246),
    bump: readU8(v, 249),
    vaultBump: readU8(v, 250),
  };
}

// ---------------------------------------------------------------------------
// UnderwriterPosition — discriminator 2, 184 bytes.
// ---------------------------------------------------------------------------
export interface UnderwriterPosition {
  pool: Pubkey;
  underwriter: Pubkey;
  deposited: bigint;
  earnedPremiums: bigint;
  lossesAbsorbed: bigint;
  depositTimestamp: bigint;
  lastClaimTimestamp: bigint;
  bump: number;
}

export function decodeUnderwriterPosition(
  data: Buffer | Uint8Array
): UnderwriterPosition {
  const v = makeReader(
    data,
    UNDERWRITER_POSITION_LEN,
    ACCOUNT_DISC_UNDERWRITER_POSITION,
    "UnderwriterPosition"
  );
  return {
    pool: readPubkey(v, 8),
    underwriter: readPubkey(v, 40),
    deposited: readU64LE(v, 72),
    earnedPremiums: readU64LE(v, 80),
    lossesAbsorbed: readU64LE(v, 88),
    depositTimestamp: readI64LE(v, 96),
    lastClaimTimestamp: readI64LE(v, 104),
    bump: readU8(v, 112),
  };
}

// ---------------------------------------------------------------------------
// Policy — discriminator 3, 320 bytes.
//
// Phase 5 F1 tail (WP-12) at offsets 216/248/250. `referrer` is the raw
// 32-byte buffer (all-zero = None sentinel); callers should rely on
// `referrerPresent === 1` instead of comparing the bytes. `referrer` is
// exposed as `Pubkey` (base58) when present, `null` when absent — the more
// ergonomic shape for downstream code.
// ---------------------------------------------------------------------------
export interface Policy {
  agent: Pubkey;
  pool: Pubkey;
  agentTokenAccount: Pubkey;
  agentId: string;
  agentIdLen: number;
  totalPremiumsPaid: bigint;
  totalClaimsReceived: bigint;
  callsCovered: bigint;
  createdAt: bigint;
  expiresAt: bigint;
  active: number;
  bump: number;
  referrer: Pubkey | null;
  referrerShareBps: number;
  referrerPresent: number;
}

export function decodePolicy(data: Buffer | Uint8Array): Policy {
  const v = makeReader(data, POLICY_LEN, ACCOUNT_DISC_POLICY, "Policy");
  const agentIdLen = readU8(v, 208);
  const referrerPresent = readU8(v, 250);
  return {
    agent: readPubkey(v, 8),
    pool: readPubkey(v, 40),
    agentTokenAccount: readPubkey(v, 72),
    agentId: readUtf8Slice(v, 104, 64, agentIdLen),
    agentIdLen,
    totalPremiumsPaid: readU64LE(v, 168),
    totalClaimsReceived: readU64LE(v, 176),
    callsCovered: readU64LE(v, 184),
    createdAt: readI64LE(v, 192),
    expiresAt: readI64LE(v, 200),
    active: readU8(v, 209),
    bump: readU8(v, 210),
    referrer: referrerPresent === 1 ? readPubkey(v, 216) : null,
    referrerShareBps: readU16LE(v, 248),
    referrerPresent,
  };
}

// ---------------------------------------------------------------------------
// Claim — discriminator 4, 288 bytes.
//
// `callId` is the 32-byte SHA-256 digest the program stores in place of the
// raw string (WP-4 addendum #9). Reconcile against `hashCallId(rawCallId)`
// from `pda.ts` if you need to confirm a specific raw id maps to a claim.
// ---------------------------------------------------------------------------
export interface Claim {
  policy: Pubkey;
  pool: Pubkey;
  agent: Pubkey;
  callId: Uint8Array;
  evidenceHash: Uint8Array;
  paymentAmount: bigint;
  refundAmount: bigint;
  callTimestamp: bigint;
  createdAt: bigint;
  resolvedAt: bigint;
  latencyMs: number;
  statusCode: number;
  triggerType: TriggerType;
  status: ClaimStatus;
  bump: number;
}

export function decodeClaim(data: Buffer | Uint8Array): Claim {
  const v = makeReader(data, CLAIM_LEN, ACCOUNT_DISC_CLAIM, "Claim");
  const triggerByte = readU8(v, 214);
  const statusByte = readU8(v, 215);
  if (triggerByte < 0 || triggerByte > 3) {
    throw new Error(`Claim: invalid trigger_type byte ${triggerByte}`);
  }
  if (statusByte < 0 || statusByte > 2) {
    throw new Error(`Claim: invalid status byte ${statusByte}`);
  }
  return {
    policy: readPubkey(v, 8),
    pool: readPubkey(v, 40),
    agent: readPubkey(v, 72),
    callId: readBytes32(v, 104),
    evidenceHash: readBytes32(v, 136),
    paymentAmount: readU64LE(v, 168),
    refundAmount: readU64LE(v, 176),
    callTimestamp: readI64LE(v, 184),
    createdAt: readI64LE(v, 192),
    resolvedAt: readI64LE(v, 200),
    latencyMs: readU32LE(v, 208),
    statusCode: readU16LE(v, 212),
    triggerType: triggerByte as TriggerType,
    status: statusByte as ClaimStatus,
    bump: readU8(v, 216),
  };
}
